### PR TITLE
Testcloud: Use qemu usermode

### PR DESCRIPTION
**Pull request by frantisekz**
Testcloud 0.5.0 is required and is available in all supported Fedoras, pypi and EPEL 8.

### CODE REVIEW

**Comment by frantisekz** [comment]
Okay, I've added latest testcloud PR and tmt with this PR applied into https://copr.fedorainfracloud.org/coprs/frantisekz/testcloud-wip/ .

However, I am still running into one issue (there might be more, I am running just some basic tmt run on top of bash rpm git clone). It seems that tmt is trying to spawn the same vm again which fails, because the port and net device id is already occupied by the first vm. I'll be very happy for any suggestions/ideas/feedback (also, I am not too tmt codebase savvy :octocat: ). @psss @thrix @martinpitt

**Comment by frantisekz** [comment]
> Okay, I've added latest testcloud PR and tmt with this PR applied into https://copr.fedorainfracloud.org/coprs/frantisekz/testcloud-wip/ .
> 
> However, I am still running into one issue...

Okay, so, this is fixed. The new version of testcloud and tmt in my copr **appear to be working just fine** and using user sessions for vms, at least in my limited testing. (The issue was that for second vm spawn in one tmt run, we already had config_data.CMD_LINE_ARGS pre-filled and appending the same netdev twice made qemu very unhappy, took me some time to debug this one).

I don't think there is a need to configure which type of session user wants for tmt, I can add that as an option if it's preferred.

Anyhow, once this gets +1, I'll merge the testcloud changes, cut a new release and push it to all the Fedora and EPEL repos, after that, this could be merged.

**Comment by frantisekz** [comment]
The other thing to watch out for/something I can add a hotfix for is that '/var/tmp/tmt/testcloud/domain-template.jinja' needs to be recreated. I can do some dumb "if xmlns not in template: burn template with fire and generate a new one" if that sounds okay?

**Comment by martinpitt** [comment]
I just re-tested the latest testcloud from the copr by itself, and [reported my results](https://pagure.io/testcloud/pull-request/83#comment-144204).

Now I removed all my chown/ssh config/etc. hacks, and tried tmt from the COPR, with the most naïve way possible:
```
git clone https://github.com/cockpit-project/cockpit-podman/
cd cockpit-podman
tmt run
```

I really like how that avoids all these problems with "I can't access /var/lib/testcloud" and [I can't test Fedor 34](https://pagure.io/testcloud/issue/84) (I locally hacked in the cloud URL in my previous runs). It downloaded and started a F34 VM (with 2 GiB RAM and 1 CPU, which is what you get on Testing Farm, so :+1:), and ran the test successfully. Everything happened in a Fedora 33 toolbox.

I also adjusted my ~/.ssh/config stanza for ssh'ing to the instance:
```
Host testcloud
    Hostname localhost
    User root
    Port 10023
    IdentityFile /var/tmp/tmt/run-001/plans/all/provision/default/id_rsa
    StrictHostKeyChecking no
    UserKnownHostsFile /dev/null
```
I usually just run one test at a time, but I suppose I could just do `ssh -p 10024 testcloud` if I ever have a second one.

That's amazing, thank you! :heart_eyes:

**Comment by martinpitt** [comment]
Next step: Learn how to enable a COPR for the test. Previously I was doing `ssh testcloud dnf -y copr enable packit/martinpitt-cockpit-13` after manually starting the VM with `testcloud` and starting the test with `tmt run --all provision --how connect -g testcloud'`. Now I tried with

    tmt run --all prepare --how install --copr packit/martinpitt-cockpit-13

but that doesn't work -- the copr does not get installed into the VM, and the packages that the plan installs are from regular Fedora. But I'll send that as a separate issue.

Thanks!

**Comment by psss** [comment]
Thanks a lot for preparing the patch! Works nicely. I just ran into an issue with the outdated `domain-template.jinja` which resulted in a traceback. As users will usually have this file already created under `/var/tmp/tmt/testcloud` I guess we should just overwrite it every time. I also think it will be better to use a dedicated key `port` in the `Guest` class. Please, review my changes in d1299d4.

**Comment by frantisekz** [comment]
> I also think it will be better to use a dedicated key `port` in the `Guest` class. Please, review my changes in [d1299d4](https://github.com/psss/tmt/commit/d1299d45cb1370b2dd20745e35606126d9fc462d).

This is definitely much better, I was making changes this way, don't know why I resorted to splitting on ":" in the end. Thanks for the improvements and review!

**Comment by psss** [comment]
I did some more experimenting and ran into an issue when waking up an already running instance:

```
> tmt run provision -h virt -i f33
/var/tmp/tmt/run-027

/plans/default
    provision
        how: virt
        image: f33
        user: root
        memory: 2048 MB
        disk: 10 GB
        progress: booting...
        distro: Fedora 33 (Cloud Edition)
        summary: 1 guest provisioned

> tmt run -l finish
/var/tmp/tmt/run-027

/plans/default
    finish
libvirt: QEMU Driver error : Doména nebyla nalezena: žádná doména nazvaná „OCfcpLEDCjfoyvhv“
Failed to stop testcloud instance: Instance doesn't exist: OCfcpLEDCjfoyvhv
```

@frantisekz, any idea what might have changed which could cause this?

**Comment by frantisekz** [comment]
> @frantisekz, any idea what might have changed which could cause this?

Yeah, it seemed bad at first, but fix was really trivial (hope it works though :) ).

**Comment by psss** [comment]
Ah, I see. Thanks for the quick fix. One note, when debugging the issue noticed this:

```
> /home/psss/git/tmt/tmt/steps/provision/testcloud.py(454)stop()
-> if self.instance:
(Pdb) self.instance.image
<testcloud.image.Image object at 0x7fc36f7099a0>
(Pdb) self.instance.image_path
'/var/tmp/tmt/testcloud/imagesONsxCFTCjFdrGIIH.qcow2'
```
Seems the `image_path` is missing a `/` after `images`. It does not affect the functionality (and is the same before the latest changes) but I wonder if that's a bug in `tmt` or `testcloud`.

**Comment by frantisekz** [comment]
> Ah, I see. Thanks for the quick fix. One note, when debugging the issue noticed this:
> 
> ```
> > /home/psss/git/tmt/tmt/steps/provision/testcloud.py(454)stop()
> -> if self.instance:
> (Pdb) self.instance.image
> <testcloud.image.Image object at 0x7fc36f7099a0>
> (Pdb) self.instance.image_path
> '/var/tmp/tmt/testcloud/imagesONsxCFTCjFdrGIIH.qcow2'
> ```
> 
> Seems the `image_path` is missing a `/` after `images`. It does not affect the functionality (and is the same before the latest changes) but I wonder if that's a bug in `tmt` or `testcloud`.

Yeah, that's a bug in testcloud, fix is in https://pagure.io/testcloud/pull-request/87 , can you try that? Thanks!

**Comment by psss** [comment]
> Yeah, that's a bug in testcloud, fix is in https://pagure.io/testcloud/pull-request/87 , can you try that? Thanks!

Confirmed, now the `image_path` contains a reasonabled (and existing) path. Thanks.

**Comment by psss** [comment]
This resolves #541 but we'll need to wait until fresh `testcloud` is released. Moving #541 to `1.5` as well.

File: tmt/steps/provision/testcloud.py
Comment (jscotka) [inline tmt/steps/provision/testcloud.py:430]: That code means, that `port` is selected randomly by testcloud library. Or It is possible to force some port mapping e.g. `22 (SSH) -> 2201(localSSHport)` of first spawned machine? Why I'm asking is, that in case I would like to connect to test machine would have port numbers more predictable.  But not sure how implementation inside testcloud look like.

**Review by jscotka — COMMENTED** [review]
In generic it seems very well and clean. 

Just few question

File: tmt/steps/provision/testcloud.py
Comment (jscotka) [inline tmt/steps/provision/testcloud.py:97]: for some multihost test, not sure how much it is implemented already in TMT, but would be possible to have there tests, that machines are able to communicate together. If using same network prefix under `user` network can communicate to each other. I think yes, but not sure.

File: tmt/steps/provision/testcloud.py
Comment (frantisekz) [reply tmt/steps/provision/testcloud.py:430]: There is a config option for testcloud: SSH_USER_PORT_BASE, which sets a start of used interval for ssh ports. It defaults to 10022 which means the first created vm will have that port, another will be 10023 and so on. It recycles ports that aren't in use, so eg. you create 3 vms, their ports will be 10022, 10023, 10024, you delete the second vm, create a new one and it'll have port 10023. I hope that answers the question.

Comment (frantisekz) [reply tmt/steps/provision/testcloud.py:430]: Thinking about this, there probably needs to be some change done to be able to change that from tmt, I can take a look at it.

File: tmt/steps/provision/testcloud.py
Comment (frantisekz) [reply tmt/steps/provision/testcloud.py:97]: I am not sure if it can be done with testcloud as it is right now, it could be achieved by implementing socket networking in testcloud imo. I think we can add it to the roadmap and implement that later as a separate PR?

File: tmt/steps/provision/testcloud.py
Comment (psss) [reply tmt/steps/provision/testcloud.py:430]: From the `tmt` point of view I would say we're set. The instance of the `Guest` class has the `port` available and can be used if necessary.

File: tmt/steps/provision/testcloud.py
Comment (psss) [reply tmt/steps/provision/testcloud.py:97]: Yes, makes sense to handle this as a separate pull request. Currently there is no support for multihost testing in `tmt`.

File: tmt/steps/provision/testcloud.py
Comment (jscotka) [reply tmt/steps/provision/testcloud.py:430]: Yep, it contains `self.port` as good solution, just when scheduled tests, sometimes would be nice to connect to the machine, and this info is not displayed in default mode, probably also not in verbose, just in debug mode probably, means, to kill the process and run it again.

File: tmt/steps/provision/testcloud.py
Comment (jscotka) [reply tmt/steps/provision/testcloud.py:97]: Yes, agreed, I was just not sure if there is any implementation for multihost in TMT right now.

File: tmt/steps/provision/testcloud.py
Comment (psss) [reply tmt/steps/provision/testcloud.py:430]: Good idea to show the `ip` and `port` to the user in verbose mode. Done in ac83c1a.

Comment (jscotka) [reply tmt/steps/provision/testcloud.py:430]: @psss BTW would be good to display full ssh command line to connect to any machine instead of just printing port and host
so that I can imagine in vebose mode something like:
```
 f"Connect via: ssh -p {self.port} -i {self.ssh_key} {self.guest}" in case not used key, display also password
```
Theoretically I as an user I do not know where are the ssh keys or if there is some randomly choosed password for any instances: minute, virtual, connect, podman, ...

Comment (psss) [reply tmt/steps/provision/testcloud.py:430]: We can add this to the debug log. But specifically for this use case `tmt run login` should provide an easy way to get to the box.

**Comment by pvalena** [comment]
I'll test on SilverBlue.

**Comment by psss** [comment]
> I'll test on SilverBlue.

@pvalena, any results from the testing? @martinpitt [tested in toolbox](https://bodhi.fedoraproject.org/updates/FEDORA-2021-fde5da6164). I think it would be nice to get this into `1.4` rather than waiting another month. Unless anybody sees any blockers.

**Comment by psss** [comment]
Squashed commits, rebased on the latest master and resolved the conflict.

**Comment by psss** [comment]
No more blocked on `testcloud` as `0.5` has already [reached](https://bodhi.fedoraproject.org/updates/?packages=testcloud) stable repos.

**Comment by pvalena** [comment]
> > I'll test on SilverBlue.
> 
> @pvalena, any results from the testing? @martinpitt [tested in toolbox](https://bodhi.fedoraproject.org/updates/FEDORA-2021-fde5da6164). I think it would be nice to get this into `1.4` rather than waiting another month. Unless anybody sees any blockers.

I was stuck on another work. I'll be able to test tmrw.

**Review by lukaszachy — COMMENTED** [review]
LGTM, however few UX issues when used with gnome-boxes:
- Random names are strange - Can we prefix them at least with tmt- (maybe tmt-runid) so it is clear they are from tmt?
- This is very likely a gnome-boxes issue but on my gnome-boxes-3.38.2-2.fc33 closing gnome-boxes kills tmt's machine, we should at least document this.
1. Run `tmt run provision -v login finish`
2. Open gnome-boxes, don't click anything
3. Close gnome-boxes
4. tmt says Connection to 127.0.0.1 closed by remote host and traceback
```
libvirt: QEMU Driver error : Requested operation is not valid: Refusing to undefine while domain managed save image exists
Traceback (most recent call last):
  File "/home/lzachar/.virtualenvs/tmt/bin/tmt", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/home/lzachar/gits/tmt/bin/tmt", line 8, in <module>
    tmt.cli.main()
  File "/home/lzachar/.virtualenvs/tmt/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/lzachar/.virtualenvs/tmt/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/lzachar/.virtualenvs/tmt/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/lzachar/.virtualenvs/tmt/lib/python3.9/site-packages/click/core.py", line 1290, in invoke
    return _process_result(rv)
  File "/home/lzachar/.virtualenvs/tmt/lib/python3.9/site-packages/click/core.py", line 1224, in _process_result
    value = ctx.invoke(self.result_callback, value, **ctx.params)
  File "/home/lzachar/.virtualenvs/tmt/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/lzachar/.virtualenvs/tmt/lib/python3.9/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/lzachar/gits/tmt/tmt/cli.py", line 295, in finito
    click_context.obj.run.go()
  File "/home/lzachar/gits/tmt/tmt/base.py", line 1154, in go
    plan.go()
  File "/home/lzachar/gits/tmt/tmt/base.py", line 644, in go
    self.finish.go()
  File "/home/lzachar/gits/tmt/tmt/steps/finish/__init__.py", line 70, in go
    guest.remove()
  File "/home/lzachar/gits/tmt/tmt/steps/provision/testcloud.py", line 475, in remove
    self.instance.remove(autostop=True)
  File "/home/lzachar/.virtualenvs/tmt/lib/python3.9/site-packages/testcloud/instance.py", line 666, in remove
    self._remove_from_libvirt()
  File "/home/lzachar/.virtualenvs/tmt/lib/python3.9/site-packages/testcloud/instance.py", line 639, in _remove_from_libvirt
    self._get_domain().undefineFlags(libvirt.VIR_DOMAIN_UNDEFINE_NVRAM)
  File "/home/lzachar/.virtualenvs/tmt/lib/python3.9/site-packages/libvirt.py", line 2938, in undefineFlags
    if ret == -1: raise libvirtError ('virDomainUndefineFlags() failed', dom=self)
libvirt.libvirtError: Requested operation is not valid: Refusing to undefine while domain managed save image exists
```

**Review by psss — APPROVED** [review]
Looks good and seems to be working as expected, confirmed by several users, no important issues or blockers identified ---> merging.

**Pull request merged**
